### PR TITLE
Start dump from last simulation's start

### PR DESCRIPTION
**Pull request by ghoto**
Added argument for search dumps to start from last simulation's start. This is helpful for not including events caused by actions such as "build", or any other action performed before the simulation.

```
 python attack_range.py -a dump -dn T1036 --last-sim
```

Without `--last-sim` flag it defaults to time defined in "dumps.yml"

### CODE REVIEW

File: modules/TerraformController.py [hunk 1]
```diff
@@ -226,6 +230,12 @@ def dump_attack_data(self, dump_name):
                     for dump in yaml.full_load(dumps):
                         if dump['enabled']:
                             dump_out = dump['out']
+                            if last_sim:
+                                # if last_sim is set, then it overrides time in dumps.yml
+                                # and starts dumping from last simulation
+                                with open("/tmp/attack-range-%s-last-sim.tmp" % self.config['range_name'], 'r') as ls:
```
Comment (miskoSplunk) [inline modules/TerraformController.py:236]: Don't know how quoting works in Python 3, but consider doing tripple quotes (""") around "/tmp/attack-range-%s-last-sim.tmp".

This is to allow that insert %s to be anything, i.e., not be prone to quoting issues.

**Review by miskoSplunk — COMMENTED** [review]
Not sure if I understood well "Added argument for search dumps to start from last simulation's start", but for me this means start logging from when current simulation starts.

To that end, to prevent users from thinking that this requires having previous simulations, could we rename this to something like "log from simulation start", or "log simulation only"?

File: modules/TerraformController.py (hunk 1 above)
Comment (ghoto) [reply modules/TerraformController.py:236]: This code is fine. There is no need for triple quotes as string is singled line. 

This does what you expect it to do (inserting a string at %s position)

**Comment by ghoto** [comment]
> Not sure if I understood well "Added argument for search dumps to start from last simulation's start", but for me this means start logging from when current simulation starts.
> 
> To that end, to prevent users from thinking that this requires having previous simulations, could we rename this to something like "log from simulation start", or "log simulation only"?

Are you referring to the description of the PR? or something in the code? Once this is approved, this description is gone. The purpose of it is so you guys understand what this feature/PR does. 

Is not clear what this feature does?

**Comment by P4T12ICK** [comment]
I think we should change the path to the attack_data path inside the project. Please use the following for the path e.g. os.path.join(os.path.dirname(__file__), '../attack_data/attack-range-%s-last-sim.tmp'). You need to specify the path relative to the path of the code file. This is needed for the detection testing service. 
What is the use case of this feature? Limiting the amount of dump data?

**Comment by ghoto** [comment]
> I think we should change the path to the attack_data path inside the project. Please use the following for the path e.g. os.path.join(os.path.dirname(**file**), '../attack_data/attack-range-%s-last-sim.tmp'). You need to specify the path relative to the path of the code file. This is needed for the detection testing service.

Yes, this was submitted prior your PR that uses this convention. This can be changed.

> What is the use case of this feature? Limiting the amount of dump data?

Yes, limit the dump data to the start of a simulation to have cleaner dumps. A use case could be:

$ run simulation T1059
$ dump T1059
$ run simulation T1030
$ dump T1030

You will end up with cleaner datasets.

**Comment by P4T12ICK** [comment]
Sounds good

**Comment by P4T12ICK** [comment]
can you do the change and then I will review it tomorrow

**Comment by P4T12ICK** [comment]
@ghoto the path needs to be still changed to be absolute in comparison to the code file. See line 258 in TerraformController.py. Which means using this one os.path.join(os.path.dirname(file), '../attack_data/attack-range-%s-last-sim.tmp').
The variable should be inside the attack_data folder.
The same for line 184 in the simulate function.
Maybe create another subfolder in attack_data to store these files.

**Comment by ghoto** [comment]
@P4T12ICK But I'm writing not to a local path. I'm writing to absolute /tmp. Perhaps you want me to use a python tmp file library.

**Review by P4T12ICK — APPROVED** [review]
looks good

File: modules/TerraformController.py (hunk 1 above)
Comment (miskoSplunk) [reply modules/TerraformController.py]: ok

**Comment by miskoSplunk** [comment]
> > Not sure if I understood well "Added argument for search dumps to start from last simulation's start", but for me this means start logging from when current simulation starts.
> > To that end, to prevent users from thinking that this requires having previous simulations, could we rename this to something like "log from simulation start", or "log simulation only"?
> 
> Are you referring to the description of the PR? or something in the code? Once this is approved, this description is gone. The purpose of it is so you guys understand what this feature/PR does.
> 
> Is not clear what this feature does?

I was thinking about potentially renaming a command option --last-sim and explaining it more clearly in code comments/docs.

If I understood this well, this change is about starting logging only after AR simulation is fully set up (vs. during the set up). Thus, using "last" may be confusing. It would indicate that there has to be a series of simulations, i.e., that there has to be another simulation proceeding the current one in order to use this --last-sim option.

**Pull request merged**